052-record-reconciled-versions.md

```diff
@@ -0,0 +1,203 @@
+
+## Proposal
+
+This proposal will cover the `Kafka` and `StrimziPodSet` custom resources, but could be extended to include other Custom Resources if there would be value in it. This proposal starts with the `Kafka` Custom Resource as this is the core component of Strimzi, and its reliance on `StrimziPodSet` means we should implement it for this CR type as well.
+
+This proposal assumes that all key value pairs mentioned will be writen to the `metadata.annotations` field [docs here](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/).
+Alternatively, there is an argument for putting them in the [metadata.labels field](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/), but given Strimzi's heavy use of the labels field, in order to not confuse a user who might think they had accidentally set the field themselves, this proposal has chosen to use annotations.
+
+
+The following are the two proposed new annotations:
+
+The `strimzi.io/reconciled` annotation is patched at the end of a **successful** reconcile, and signals to a user that the operator at version `X` was able to reconcile to successful completion.
+This would be the annotation a user or System Test would query against to check whether upgrade was complete.
+```
+strimzi.io/reconciled: '0.37.0'
+```
+
+The `strimzi.io/reconciling` annotation is patched to the CR at the start of reconcile and signals to a user that the operator is attempting to reconcile the CR at version Y.
+```
+strimzi.io/reconciling: '0.37.0'
+```
+This annotation is required for the following reasons:
+- It signals that a reconcile on version Y has started, and that the operator is functional without checking the logs of the operator.
+- The operator updates the `strimzi.io/reconciling` annotation value on reconciliation startup without overriding the `strimzi.io/reconciled` annotation. If the operator errors during reconcile it will be clear that the currently reconciled version does not match the version that the operator is applying. A two stage commit mechanism here gives much more information regarding whether it started the reconciliation versus whether it finished successfully.
+
+An important note here, in order to avoid unnecessary extra reconcilliations, and to avoid constant re-adding and removal of the annotation, this proposal suggests that `strimzi.io/reconciling` remains, even after a reconcile, and that the annotation is only ever updated once a new operator version attempts the reconcile.
+
+
+The following two sections assume that along with status updates, the operator now also patches the annotations alongside to reflect the new annotation state, which will require a patch/update to the CR, rather than a simple `updateStatus` call.
```

Comment:
TBH, this seems to overcomplicate this quite significantly. It also significantly increases the number of API calls. 

Have you considered to:
* Use just one annotation
* Roll it out to StrimziPodSets and Pods as part of their regular reconciliation?

That should give you exactly the same information about the progress with significantly less effort: 
* You will know how far did the new version of the operator got by looking at where the new annotations are:
  * Did it reconciled the StrimziPodSet? If yes, it has the new version. If not, it has the old version
  * Did it managed to roll out the pods? If yes, the pods will have the new versions. If not, it still has the old version

This is also much more accurate information. Because if the operator is failing in the middle, you will know exactly how far it got. If you do the annotations on the beginning and at the end, you do not have the information about how far did it got only that it is _in the process_. But this way, you can from the resources know exactly that it is stuck for example _in between Kafka and ZooKeeper_.